dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/TrackerBundleParamsConverter.java

```diff
@@ -0,0 +1,182 @@
+ *  |___ENROLLMENT 1, ENROLLMENT 2
+ *  |
+ *  |___EVENT 1, EVENT 2, EVENT 3, EVENT 4
+ *
+ * </pre>
+ *
+ * This converter also assigns UIDs to Tracked Entities, Enrollment and Events if the payload does not contain UIDs
+ *
+ * @author Luciano Fiandesio
+ */
+public class TrackerBundleParamsConverter
+    extends
+    StdConverter<TrackerBundleParams, TrackerBundleParams>
+{
+    @Override
+    public TrackerBundleParams convert( TrackerBundleParams bundle )
+    {
+        if ( hasNestedStructure( bundle ) )
+        {
+            // pre assign UIDs to entities, if UIDs are missing
+            generateUid( bundle );
+
+            flattenPayload( bundle );
+        }
+
+        return bundle;
+    }
+
+    private void flattenPayload( TrackerBundleParams bundle ) {
+
+        List<Event> events = new ArrayList<>();
+
+        List<Enrollment> enrollments = new ArrayList<>();
+
+        for ( Enrollment enrollment : bundle.getTrackedEntities().stream()
+                .flatMap( l -> l.getEnrollments().stream() ).collect( Collectors.toList() ) )
+        {
+            events.addAll( enrollment.getEvents().stream().map( e -> addParent( e, enrollment.getEnrollment() ) )
+                .collect( Collectors.toList() ) );
+
+            // remove events from enrollment
+            enrollment.setEvents( Collections.emptyList() );
+        }
+
+        for ( TrackedEntity trackedEntity : bundle.getTrackedEntities() )
+        {
+            enrollments.addAll( trackedEntity.getEnrollments().stream()
+                .map( e -> addParent( e, trackedEntity.getTrackedEntity() ) ).collect( Collectors.toList() ) );
+
+            // remove enrollments from tracked entities
+            trackedEntity.setEnrollments( Collections.emptyList() );
+        }
+
+        bundle.getEvents().addAll( events );
+        bundle.getEnrollments().addAll( enrollments );
+    }
+
+    /**
+     * Make sure that the Enrollment has the parent ID correctly set
+     *
+     */
+    private Enrollment addParent( Enrollment enrollment, String trackerEntityId )
+    {
+        if ( StringUtils.isEmpty( enrollment.getTrackedEntity() ) )
+        {
+            enrollment.setTrackedEntity( trackerEntityId );
+        }
+        return enrollment;
+    }
+
+    /**
+     * Make sure that the Event has the parent ID correctly set
+     *
+     */
+    private Event addParent( Event event, String enrollmentId )
+    {
+
+        if ( StringUtils.isEmpty( event.getEnrollment() ) )
+        {
+            event.setEnrollment( enrollmentId );
+        }
+        return event;
+    }
+
+    /**
+     * Check if the {@see TrackerBundleParams} has a nested structure
+     *
+     * @param bundle a {@see TrackerBundleParams}
+     * @return true, if TEIs have nested enrollments
+     */
+    private boolean hasNestedStructure( TrackerBundleParams bundle )
+    {
+        return bundle.getEnrollments().isEmpty();
+    }
+
+    private void generateUid( TrackerBundleParams params )
```

Comment:
If we have it here.. should we remove it from other places? its a small performance opt... but no need if we are doing it before anyways...